File: Linked List/C++/3. Reverse a Linked List.cpp
/* Problem Link:
   https://leetcode.com/problems/reverse-linked-list/

   Ques. 
   Given the head of a singly linked list, reverse the list, and return the reversed list.

  Explaination:
  Reversing a linked list means reversing the elements present in linked list, i.e. the 
  linked list after reversing should display the elements in reverse order as they were 
  entered. 
  Now, reversing the elements in linked list can have two approaches- either reversing all 
  the elements or to reverse the links and the latter is more logical.
*/



/*
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

#include<iostream>
using namespace std;

class Solution {
    
    public: 
    ListNode* reverseList(ListNode* head) {         //code to reverse the linked list by reversing nodes
      
       ListNode *p =head, *q = NULL, *r = NULL;

        while(p!=NULL){              //continue the loop until p reaches NULL 

            r=q;              //make pointer r points to the place where pointer q is pointing 
            q=p;              //make pointer q points to the place where pointer p is pointing 
            p=p->next;         //move pointer p to point to next element

            q->next=r;        //changing the linking of nodes to reverse the linked list
        }
        head=q;             //making head points on the last node of given linked list as in revsere it becomes the first node
       
        return head;       // return the reverse linked list
    }
};

/********************* Time and Space complexity *************************************/


/* Time Complexity of above method: O(n)
   Space Complexity of above method: O(1)  */


/****************************** Testcases:  ******************************************/

/*
   Testcase 1-
   Input: head = [1,2,3,4,5]
   Output: [5,4,3,2,1]

   Testcase 2-
   Input: head = [1,2]
   Output: [2,1]

   Testcase 3-
   Input: head = []
   Output: []
*/

File: Linked List/Java/3_Reverse_a_Linked_List.java
// Question  https://leetcode.com/problems/reverse-linked-list/

import java.util.Scanner;

class LinkedList {
    ListNode head; // head of list

    /* Linked list Node*/
    static class ListNode {
        int data;
        ListNode next;
        ListNode(int data)
        {
            this.data = data;
            next = null;
        }
    }

    public static ListNode reverseList(ListNode head) {
        if(head == null){
            return null;
        }
        ListNode temp = head, after = null, prev = null;
        while(temp != null){
            after = temp.next;
            temp.next = prev;
            prev = temp;
            temp = after;
        }
        head = prev;
        return head;
    }

    public void push(int new_data)
    {
        ListNode new_node = new ListNode(new_data);
        ListNode temp = head;

        while(temp.next != null)
        {
            temp = temp.next;
        }

        temp.next = new_node;
    }

    public void printList()
    {
        ListNode tnode = head;
        while (tnode != null) {
            System.out.print(tnode.data + " ");
            tnode = tnode.next;
        }
    }

    public static void main(String[] args)
    {
        Scanner sc =new Scanner(System.in);

        LinkedList llist = new LinkedList();

        System.out.println("\nEnter Linked list size:");
        int n = sc.nextInt();

        if(llist.head == null)
        {
            int e = sc.nextInt();
            ListNode new_node =new ListNode(e);
            llist.head = new_node;
        }

        for(int i = 1;i < n;i ++)
        {
            int e = sc.nextInt();
            llist.push(e);
        }

        System.out.println("\nCreated Linked list is:");
        llist.printList();

        System.out.println("\nReverse of Linked list is:");

        ListNode res = reverseList(llist.head);

        while(res!=null){
            System.out.print( res.data + " ");
            res = res.next;
        }
    }
}

// Examples
/**
 * Example 1:
 * Input: head = [1,2,3,4,5]
 * Output: [5,4,3,2,1]
 *
 * Example 2:
 * Input: head = [1,2]
 * Output: [2,1]
 */
